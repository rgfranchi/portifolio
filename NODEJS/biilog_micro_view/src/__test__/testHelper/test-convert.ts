export const convertIdToNumeric = (data: any) => {
  if (Array.isArray(data)) {
    data.forEach((element) => {
      element = convertIdToNumeric(element);
    });
  }

  if (typeof data === "object" && data) {
    Object.entries(data).forEach((value, key) => {
      const k = value[0];
      const v = value[1];
      if (typeof v === "object") {
        data[k] = convertIdToNumeric(v);
      }
      if (k === "id") {
        const parse = parseInt(data[k]);
        if (!isNaN(parse)) {
          data[k] = parse;
        }
      }
    });
  }
  return data;
};

export const convertDateYYYYMMDDToDD_MM_YYYY = (dateSrt: string) => {
  const dArr = dateSrt.split("-"); // ex input "2010-01-18"
  return dArr[2] + "/" + dArr[1] + "/" + dArr[0]; //ex out: "18/01/2010"
};
