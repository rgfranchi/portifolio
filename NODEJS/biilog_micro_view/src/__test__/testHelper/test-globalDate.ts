/**
 * Fixa a data e hora do sistema para teste.
 * Realizar chamada no beforeAll(() => {});
 * Criar variÃ¡vel global : const realDate = Date.bind(global.Date);
 * E incluir no afterAll(() => {global.Date = realDate;});
 * @param YYYY
 * @param MM
 * @param DD
 * @param HH
 * @param mm
 * @param ss
 */

export const fixedGlobalDate = (
  YYYY: number,
  MM: number,
  DD: number,
  HH: number,
  mm: number,
  ss: number
) => {
  // https://github.com/facebook/jest/issues/2234#issuecomment-573621597
  const fixedDate = new Date(YYYY, MM, DD, HH, mm, ss);
  const d = Date;
  // This will only mock any Date objects instantiated with new
  // and not Date.now().
  const mockGlobal = global;
  // @ts-ignore
  mockGlobal.Date = jest.fn(() => fixedDate);
  mockGlobal.Date.parse = d.parse;
  mockGlobal.Date.UTC = d.UTC;
  mockGlobal.Date.now = d.now;
};
