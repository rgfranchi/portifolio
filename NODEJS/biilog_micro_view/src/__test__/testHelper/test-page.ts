import toJson from "enzyme-to-json";

/**
 * verifica o conteúdo de construção dos componentes Page.
 * @param {*} values -> array defaultPageText de test-text-defaults.js
 * @param {*} warpper -> componente que será testado, verifica Snapshot
 */
export const verifyPageContents = (values: any, wrapper: any) => {
  values.find.forEach((search: any) => {
    // console.log(search);
    expect(wrapper.find(search).exists()).toBeTruthy();
  });
  // console.log(toJson(wrapper));
  expect(toJson(wrapper)).toMatchSnapshot();
};

/**
 * verifica o conteúdo de construção dos componentes Page.
 * @param {*} values -> array defaultPageText de test-text-defaults.js
 * @param {*} warpper -> componente que será testado, verifica Snapshot
 */
export const verifyNotAccess = (values: any, wrapper: any) => {
  // console.log(values.find);
  values.find.forEach((search: any) => {
    // console.log(search);
    expect(wrapper.find(search).exists()).toBeFalsy();
  });
};

/**
 * verifica o conteúdo de construção dos componentes Page.
 * @param {*} values -> array defaultPageText de test-text-defaults.js
 * @param {*} history -> histórico da url para validação.
 */
export const loadHistoryPageContents = (values: any, history: any) => {
  history.entries[0].key = values.historyKey1;
  if (values.url1 !== "") {
    history.push(values.url1);
    history.entries[1].key = values.historyKey2;
  }
};
