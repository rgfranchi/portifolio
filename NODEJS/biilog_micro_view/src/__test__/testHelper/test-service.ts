import {
  screen,
  waitForElement,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { confirmModal } from "../../redux/modalSlice";

/**
 * Verifica se os elementos de save da tela foram executados com sucesso.
 * @param {boolean} isSuccess
 * @param {*} store
 */
export const verifyScreenAfterSave = async (isSuccess: boolean, store: any) => {
  // verifica se o modal Loading está ativo
  expect(screen.getByText(/Loading/i)).toBeInTheDocument();
  // verifica mensagem de falha.
  expect(store.getState().alert.hidden).toBe(true);
  // aguarda modal finalizar
  await waitForElementToBeRemoved(() => screen.getByText(/Loading/i));

  // Verifica estado dos elementos.
  if (isSuccess) {
    expect(store.getState().modals.success.show).toBe(true);
    expect(store.getState().modals.success.textButton).toMatch(/Retornar/i);
  } else {
    expect(store.getState().alert.hidden).toBe(false);
    expect(store.getState().alert.message).toMatch(/ERROR STATUS\[/i);
  }
};

/**
 * Executa comando de exclusão da tela List
 * @param {number} buttonPosition
 * @param {*} store
 */
export const verifyExecuteScreenDelete = async (
  buttonPosition: number,
  store: any
) => {
  // clica sobre o segundo link
  userEvent.click(screen.getAllByRole("button")[buttonPosition]);
  // verifica se alterou para exibir modal
  expect(store.getState().modals.confirm.show).toBe(true);
  // variável de NÃO executado.
  expect(store.getState().modals.confirm.executed).toBe(0);
  // realiza chamada de execução do modal.
  store.dispatch(confirmModal({ show: true, executed: 1 }));
  expect(store.getState().modals.confirm.executed).toBe(1);

  await waitForElement(() => screen.getByText(/Carregando/i));

  await waitForElementToBeRemoved(() => screen.getByText(/Carregando/i));
};

/**
 * Verifica e guarda carregar valores no tela, Create e Update
 * @param {boolean} isSuccess
 * @param {"modal" | "alert"} type "modal" | "alert"
 * @param {*} store
 */
export const loadingScreenValues = async (
  isSuccess: boolean,
  type: "modal" | "alert",
  store: any
) => {
  // state do loading exibido
  expect(store.getState().modals.loading.show).toBe(true);
  // elementos carregando Página de Loading
  expect(screen.getByText(/Carregando/i)).toBeInTheDocument();
  // verifica mensagem de falha.
  expect(store.getState().alert.hidden).toBe(true);
  // aguarda loading finalizar.
  await waitForElementToBeRemoved(() => screen.getByText(/Carregando/i));
  if (type === "modal") {
    if (isSuccess) {
      // state do loading oculto
      expect(store.getState().modals.loading.show).toBe(false);
    } else {
      expect(store.getState().modals.fail.show).toBe(true);
      expect(store.getState().modals.fail.textFail).toMatch(/vazia/i);
    }
  } else {
    expect(store.getState().alert.hidden).toBe(false);
    expect(store.getState().alert.message).toMatch(/vazia/i);
  }
};
