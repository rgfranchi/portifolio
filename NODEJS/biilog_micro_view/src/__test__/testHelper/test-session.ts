import { listGroupMock } from "../server/serverGroup";
import {
  ISessionAccess,
  ISessionSlice,
} from "../../interfaces/SessionInterface";

/**
 * Todo: Utilizar interface :ISessionSlice
 */
export const objMockEmptySession: ISessionSlice = {
  userId: 0,
  company: { id: 0, nome: "" },
  group: { id: 0, nome: "", access: [] },
};

export const objMockSession: ISessionSlice = {
  userId: 3,
  company: { id: 2, nome: "WINE SESSION" },
  group: listGroupMock[0],
};

export const objMockMenuAccess: ISessionAccess[] = [
  {
    path: "/company",
    page: {
      create: {
        label: "Nova Empresa",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/operator",
    page: {
      create: {
        label: "Inserir Operador",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/user",
    page: {
      create: {
        label: "Novo Usuário",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/machine",
    page: {
      create: {
        label: "Nova Máquina",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/forklift",
    page: {
      create: {
        label: "Inserir Equipamento",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
];
