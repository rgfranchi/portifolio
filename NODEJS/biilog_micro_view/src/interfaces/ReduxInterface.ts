import { ISessionSlice } from "./SessionInterface";
import {
  IConfirmModal,
  IFailModal,
  ILoadingModal,
  ISuccessModal,
} from "./ModalsInterface";

export interface IState {
  companies: ICompanySlice[] | [];
  session: ISessionSlice;
  modals: IModalSlice;
  alert: IAlertSlice;
}

export interface IReducer {
  (): IState;
}

export interface IReduxAction<T> {
  type: string;
  payload?: T;
}

export interface IReduxDispatch<T> {
  (value: IReduxAction<T>): void;
}

export interface IModalSlice {
  loading: ILoadingModal;
  confirm: IConfirmModal;
  fail: IFailModal;
  success: ISuccessModal;
}

export interface IAlertSlice {
  hidden: boolean;
  message: string | null;
  variant?:
    | "primary"
    | "secondary"
    | "success"
    | "danger"
    | "warning"
    | "info"
    | "light"
    | "dark"
    | string;
}

export interface ICompanySlice {
  id: number;
  nome: string;
}
