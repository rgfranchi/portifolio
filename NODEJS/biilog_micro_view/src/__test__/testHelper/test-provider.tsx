import React from "react";
import { MemoryRouter, Route, Router } from "react-router-dom";

import { objMockSession } from "./test-session";

import { Provider } from "react-redux";

//componentes de tela.
import { ModalLoading } from "../../components/ModalsComponent";

/**
 * Renderiza valores para teste.
 * Utilizando o render do framework @testing-library/react
 * Componente ModalLoading.
 * @param {*} render
 * @param {*} ui
 * @param {*} history
 * @param {*} store
 * @returns
 */
export const renderProviderRouter = (
  render: any,
  ui: any,
  history: any,
  store: any,
  modal: boolean = true
) => {
  // @ts-ignore
  function Wrapper({ children }) {
    return providerRouter(children, history, store, modal);
  }
  // @ts-ignore
  return render(ui, { wrapper: Wrapper });
};

/**
 * ConstrÃ³i com base no renderProviderRouter incluindo MemoryRouter e Router.
 * @param {*} render
 * @param {*} ui
 * @param {*} history
 * @param {*} store
 * @param {*} initialEntries
 * @param {*} path
 * @returns
 */
export const renderProviderMemoryRouter = (
  render: any,
  ui: any,
  history: any,
  store: any,
  initialEntries: any,
  path: string,
  modal: boolean = true
) => {
  const wrapper = (
    <MemoryRouter initialEntries={[initialEntries]}>
      <Route path={path}>{ui}</Route>
    </MemoryRouter>
  );
  return renderProviderRouter(render, wrapper, history, store, modal);
};

/**
 * Cria as funcionalidades do Provider e Router
 * Utilizando o mount do framework Enzyme
 * @param {*} mount
 * @param {*} children
 * @param {*} history
 * @param {*} store
 * @param {boolean} modal
 * @returns
 */
export const mountProviderRouter = async (
  mount: any,
  children: any,
  history: any,
  store: any,
  modal: boolean = true
) => {
  return await mount(providerRouter(children, history, store, modal));
};

export const createSessionCurrentUser = (store: any) => {
  store.dispatch({
    type: "session/currentUser",
    payload: objMockSession,
  });
};

const providerRouter = (
  children: any,
  history: any,
  store: any,
  modal: boolean
) => {
  return (
    <Provider store={store}>
      <Router history={history}>
        {modal && <ModalLoading />}
        {children}
      </Router>
    </Provider>
  );
};
