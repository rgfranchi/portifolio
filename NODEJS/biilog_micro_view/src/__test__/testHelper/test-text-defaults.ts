import { IAlertSlice, ICompanySlice } from "../../interfaces/ReduxInterface";

export let defaultMessageTextData = {
  vaziaResponse: /vazia/i,
  falhaAcessarServidor: /Falha ao acessar servidor/i,
  idNaoDefinido: /ID não definido/i,
  criadoValor: "VALOR CRIADO XPTO",
  criadoSucesso: /Criado com sucesso/i,
  atualizadoValor: "VALOR ATUALIZADO XPTO",
  atualizadoSucesso: /Atualizado com sucesso/i,
  excluidoSucesso: "DELETED",
  erro208: "RETORNA_ERRO_208", // resposta Valida porem não esperada.
  erro208Mensagem: /INVALIDO Status/i,
  erro400: "RETORNA_ERRO_400", // resposta Valida porem não esperada.
  erro400Mensagem: /ERROR STATUS/i,
};

export let defaultFormTextValidate = {
  preenchimento: /Preenchimento obrigatório/i,
  minimo3caracteres: /com no mínimo 3 caracteres/i,
  emailInvalido: /e-mail inválido/i,
  cepFormato: /formato de 99999-999/i,
  cnpjInvalido: /CNPJ inválido/i,
  selecioneOpcao: /Selecione Uma Opção/i,
};

export let defaultPageText = {
  list: {
    access: { create: { path: "/new" }, update: {}, delete: {} },
    historyKey1: "KEY_LIST1",
    url1: "",
    historyKey2: "",
    find: ["TitleComponent", "ButtonCreate", "List"],
  },
  create: {
    access: { create: { path: "/new" }, update: {}, delete: {} },
    historyKey1: "KEY_NEW1",
    url1: "//new",
    historyKey2: "KEY_NEW2",
    find: ["TitleComponent", "CreateForm"],
  },
  update: {
    access: { create: { path: "/new" }, update: {}, delete: {} },
    historyKey1: "KEY_UPDATE1",
    url1: "//update/2",
    historyKey2: "KEY_UPDATE2",
    find: ["TitleComponent", "UpdateForm"],
  },
};

export let alertSliceDefault: IAlertSlice = {
  hidden: true,
  message: null,
  variant: "primary",
};

export let companySliceDefault: ICompanySlice = {
  id: 3,
  nome: "Company Slice Mock",
};

// @ts-ignore
export let modalSliceDefault: IConfirmModal = {
  loading: {
    show: false,
    message: null,
  },
  confirm: {
    show: false,
    executed: 0, // incrementa a cada nova tentativa.
    disabledButtons: false,
    textTitle: null,
    textBody: null,
    textAction: "Executar",
    errorExecute: null,
    variantAction: "primary",
  },
  fail: {
    show: false,
    redirect: null,
    textFail: "Falha ao Executar",
    textButton: "Falhou ... ",
  },
  success: {
    show: false,
    redirect: null,
    textSuccess: "Sucesso!!!",
    textButton: "OK",
  },
};
