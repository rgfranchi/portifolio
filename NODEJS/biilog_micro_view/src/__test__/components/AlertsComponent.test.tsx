import React from "react";

import { createMemoryHistory } from "history";

import { useDispatch } from "react-redux";
import { cleanup, render, fireEvent } from "@testing-library/react";

import store from "../../redux/store";
import { listAlert } from "../../redux/alertSlice";

import { AlertList } from "../../components/AlertsComponent";
import { IAlertSlice } from "../../interfaces/ReduxInterface";
import { renderProviderRouter } from "../testHelper/test-provider";

// objeto que é enviado
let alertDispatch: IAlertSlice;
let history: any = null;
beforeEach(() => {
  cleanup();
  history = createMemoryHistory();
});

describe("AlertList action", () => {
  let ret: any = null;
  beforeEach(() => {
    ret = renderSlice(false, "Mensagem teste", "info");
  });
  test("snapshot", () => {
    expect(ret.container).toMatchSnapshot();
  });
  test("open close", () => {
    expect(ret.getByText(/×/i).closest("div")).not.toHaveAttribute("hidden");
    expect(ret.queryAllByText(/Mensagem teste/).length).toBe(1);
    fireEvent.click(ret.getByText(/×/i));
    expect(ret.getByText(/×/i).closest("div")).toHaveAttribute("hidden");
  });
});

describe("AlertList parameters", () => {
  test("default", () => {
    const ret = renderSlice(false);
    expect(ret.queryAllByText(/não definida/).length).toBe(1);
  });

  test("paramentos mensagem:objeto", () => {
    const ret = renderSlice(
      false,
      {
        nome: "teste nome",
        id: 999,
        objeto: { id: 333, url: "http://1234.com.br" },
      },
      "danger"
    );
    const res = ret.getByText("id:999").parentElement?.textContent;
    expect(res).toContain("nome:teste nome");
    expect(res).toContain("id:333");
    expect(res).toContain("url:http://1234.com.br");
  });
});

const MockDispatchListAlert: React.FC = () => {
  const dispatch = useDispatch();
  dispatch(listAlert(alertDispatch));
  return <></>;
};

const renderSlice = (hidden: boolean, message?: any, variant?: string) => {
  alertDispatch = {
    hidden: hidden,
    message: message,
    variant: variant,
  };
  return renderProviderRouter(
    render,
    <>
      <MockDispatchListAlert />
      <AlertList />
    </>,
    history,
    store
  );
};
