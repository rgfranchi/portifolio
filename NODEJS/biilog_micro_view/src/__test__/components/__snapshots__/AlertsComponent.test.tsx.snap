// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertList action snapshot 1`] = `
<div>
  <div
    class="fade alert alert-info alert-dismissible show"
    role="alert"
  >
    <button
      class="close"
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
      <span
        class="sr-only"
      >
        Close alert
      </span>
    </button>
    <ul>
      <li>
        Mensagem teste
      </li>
    </ul>
  </div>
</div>
`;
